### PR TITLE
Translate Macros That Use Runtime Identifiers To Inline Functions

### PR DESCRIPTION
**Pull request by freakmangd**
This PR changes how macros deemed to use runtime identifiers are generated. The current behavior generates decls that throw a compiler error upon being referenced:
`test.h`
```c
extern float foo;
#define FOO_TWICE foo * 2.0f
#define FOO_NEGATIVE -foo
#define BAR 10.0f
#define BAR_TWICE BAR * 2.0f
```
`zig translate-c test.h`
```zig
pub extern var foo: f32;
pub const FOO_TWICE = foo * @as(f32, 2.0);
pub const FOO_NEGATIVE = -foo;
pub const BAR = @as(f32, 10.0);
pub const BAR_TWICE = BAR * @as(f32, 2.0);
```
Referencing `FOO_TWICE` or `FOO_NEGATIVE` in your code will lead to a compile error:
`main.zig`
```zig
const std = @import("std");
const c = @cImport(@cInclude("test.h"));

export const foo: f32 = 10;

pub fn main() void {
    _ = c.FOO_TWICE;
}
```
```
~/zig/_test/.zig-cache/o/333d83848f7c681a55f06f92305dabbc/cimport.zig:499:27: error: unable to evaluate comptime expression
pub const FOO_TWICE = foo * @as(f32, 2.0);
                      ~~~~^~~~~~~~~~~~~~~
~/zig/_test/.zig-cache/o/333d83848f7c681a55f06f92305dabbc/cimport.zig:499:23: note: operation is runtime due to this operand
pub const FOO_TWICE = foo * @as(f32, 2.0);
                      ^~~
```
```
~/zig/_test/.zig-cache/o/333d83848f7c681a55f06f92305dabbc/cimport.zig:500:26: error: unable to evaluate comptime expression
pub const FOO_NEGATIVE = -foo;
                         ^~~~
```

With this PR, this code is generated:
```zig
pub extern var foo: f32;
pub inline fn FOO_TWICE() @TypeOf(foo * @as(f32, 2.0)) {
    return foo * @as(f32, 2.0);
}
pub inline fn FOO_NEGATIVE() @TypeOf(-foo) {
    return -foo;
}
pub const BAR = @as(f32, 10.0);
pub const BAR_TWICE = BAR * @as(f32, 2.0);
```
`main.zig` now runs, and if we print the result of calling `FOO_TWICE` we get the expected result:
```zig
std.log.info("{d}", .{c.FOO_TWICE()});
```
```
info: 20
```
The goal is to only change the behavior of previously unreferenceable macros.

### CODE REVIEW

**Comment by f-cozzocrea** [comment]
I don't really like that a macro would be translated into either a value or a function, but this is probably the only way to successfully translate these. The example macros are basically inline functions.

I think a `warn()` should be added to explain the translation to users. Something like:
"Macro "{s}" translated to a function because it contains a runtime value."

**Comment by Vexu** [comment]
Instead of traversing the AST you could make `parseCPrimaryExpr` set a flag in `MacroCtx` when it encounters a `var` reference and then convert it to an inline function at the end of `transMacroDefine`.

A similar system could then later be used to fix related issue #17862

> I think a `warn()` should be added to explain the translation to users. Something like:
> "Macro "{s}" translated to a function because it contains a runtime value."

I'd also like you to add this.

**Comment by freakmangd** [comment]
Hi! Thank you both for your feedback, I've implemented both things. However, I couldn't find a way to get the `warn` right next to or inside the function declaration without breaking the tests.

**Comment by f-cozzocrea** [comment]
If you add the generated warning to the expected output, that should fix the test failures. You may also need to add some blank lines between non-consecutive outputs.

File: src/translate_c.zig [hunk 1]
```diff
@@ -5768,6 +5793,11 @@ fn parseCPrimaryExpr(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!Node {
             if (builtin_typedef_map.get(mangled_name)) |ty| return Tag.type.create(c.arena, ty);
             const identifier = try Tag.identifier.create(c.arena, mangled_name);
             scope.skipVariableDiscard(identifier.castTag(.identifier).?.data);
+            refs_var: {
+                const ident_node = c.global_scope.sym_table.get(slice) orelse break :refs_var;
+                const var_decl_node = ident_node.castTag(.var_decl) orelse break :refs_var;
+                m.refs_var_decl = !var_decl_node.data.is_const;
```
Comment (Vexu) [inline src/translate_c.zig]: Slight problem here:
```c
int a = 0;
const int b = 1;

#define foo a + b
```
```zig
pub export var a: c_int = 0;
pub export const b: c_int = 1;

pub const foo = a + b;
```

```suggestion
                if (!var_decl_node.data.is_const) m.refs_var_decl = true;
```

**Review by Vexu — APPROVED** [review]
> However, I couldn't find a way to get the warn right next to or inside the function declaration without breaking the tests.

Ah right, `warn` wasn't really made for use with macros so the comment is getting added directly to the scope and macros are first added to a table.

Removing `macro_table` and directly `addTopLevelDecl`ing does seem to work so I guess it isn't needed anymore.

File: src/translate_c.zig (hunk 1 above)
Comment (freakmangd) [reply src/translate_c.zig]: Nice catch!

**Pull request merged**